spec/models/spree/line_item_spec.rb

```diff
@@ -40,10 +40,13 @@ module Spree
 
       describe "finding line items with and without tax" do
         let(:tax_rate) { create(:tax_rate, calculator: Spree::Calculator::DefaultTax.new) }
-        let!(:adjustment1) { create(:adjustment, adjustable: li1, originator: tax_rate, label: "TR", amount: 123, included_tax: 10.00) }
-        let!(:adjustment2) { create(:adjustment, adjustable: li1, originator: tax_rate, label: "TR", amount: 123, included_tax: 10.00) }
+        let!(:adjustment1) { create(:adjustment, originator: tax_rate, label: "TR", amount: 123, included_tax: 10.00) }
 
-        before { li1; li2 }
+        before do
+            li1
+            li2
+            li1.adjustments << adjustment1
```

Comment:
I think something like this is causing #4898, it looks like there's a `Spree::Calculator` being created somewhere as part of creating an associated adjustment, but it isn't being saved or associated properly in some way.

I'm wondering if there's something that needs changing in the code instead of just the specs? Like here we're having to manually associate the adjustment to the line item to fix the spec, but does this indicate that adjustments will not be being associated correctly when the app is running?

Reply:
yeah, I thought exactly the same.
I have found #4890 is also caused by some database cleaner issue that could be caused by something related to this, I agree.

ping @mkllnk @sauloperez thoughts?

Reply:
The bits we manually fixed in relation to enterprise_roles were similar, no?

Reply:
Interesting... I just tried bumping the `database_cleaner` version with `bundle update database_cleaner`. It moved from 1.7.8 to 1.8.3 and those failing specs have changed slightly; there's a warning output during those tests: `WARNING:  there is already a transaction in progress`, and the error messages are different...

Reply:
I think after each test it closes all open transactions, and somehow there are some incomplete transactions left open due to associations not saving or some error being raised and then rescued in an odd way...

Reply:
Possibly related: https://github.com/DatabaseCleaner/database_cleaner/issues/198

Reply:
Also possibly related: https://github.com/rails/rails/pull/14488

Reply:
Maybe @jonleighton has some advice, as he's all over the code review in the above mentioned Rails PR from 2014 :point_up: 

:pray:

Reply:
![](https://media.giphy.com/media/l1AsUvwEEBorHpMpG/giphy.gif)

Reply:
Hi! I've had a little look but I'm afraid it's impossible for me to say anything useful without spending a lot more time digging. I don't have enough context at the moment to understand this issue, or whether it's even related to the Rails PR you linked (which was a long time ago and my memory is pretty hazy!) Sorry I can't be of more help.

Reply:
Haha, no worries.

Reply:
Notes/discussion moved to #4898. I think we've almost got it cracked...

Reply:
So, this PR is still needed after the fix to #4890 and #4898.
I'd say we will be ok with this, it happens in factory code but in live code this is not how  adjustments are assigned to line items. Adjustments are created in  
Spree::Core::CalculatedAdjustments through target.create_adjustments which would break a lot of specs if it wasn't working.
I'd merge this and move forward.